sw/host/opentitanlib/src/util/bigint.rs

```diff
@@ -0,0 +1,375 @@
+        res.check_len()
+    }
+
+    /// Returns the bit length.
+    ///
+    /// Bit length of `FixedSizeBigInt` is the minimum number of bits required to represent its
+    /// value. The underlying storage may be larger.
+    pub(crate) fn bit_len(&self) -> usize {
+        self.0.bits()
+    }
+
+    /// Returns the byte representation in little-endian order.
+    pub(crate) fn to_le_bytes(&self) -> Vec<u8> {
+        self.0.to_bytes_le()
+    }
+
+    /// Returns the byte representation in big-endian order.
+    pub(crate) fn to_be_bytes(&self) -> Vec<u8> {
+        self.0.to_bytes_be()
+    }
+}
+
+impl<const BIT_LEN: usize, const EXACT_LEN: bool> From<FixedSizeBigInt<BIT_LEN, EXACT_LEN>>
+    for BigUint
+{
+    fn from(bigint: FixedSizeBigInt<BIT_LEN, EXACT_LEN>) -> BigUint {
+        bigint.0
+    }
+}
+
+impl<const BIT_LEN: usize, const EXACT_LEN: bool> ParseInt for FixedSizeBigInt<BIT_LEN, EXACT_LEN> {
+    type FromStrRadixErr = ParseBigIntError;
+
+    fn from_str_radix(src: &str, radix: u32) -> Result<Self, Self::FromStrRadixErr> {
+        let res =
+            Self(BigUint::from_str_radix(src, radix).map_err(ParseBigIntError::ParseBigIntError)?);
+        res.check_len()
+    }
+}
+
+impl<const BIT_LEN: usize, const EXACT_LEN: bool> fmt::Display
+    for FixedSizeBigInt<BIT_LEN, EXACT_LEN>
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let s = format!("{:#0width$x}", self.0, width = Self::BYTE_LEN * 2 + 2);
+        f.pad_integral(true, "", &s)
+    }
+}
+
+impl<const BIT_LEN: usize, const EXACT_LEN: bool> fmt::LowerHex
+    for FixedSizeBigInt<BIT_LEN, EXACT_LEN>
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.pad_integral(true, "0x", &self.0.to_str_radix(16))
+    }
+}
+
+impl<const BIT_LEN: usize, const EXACT_LEN: bool> fmt::UpperHex
+    for FixedSizeBigInt<BIT_LEN, EXACT_LEN>
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut s = self.0.to_str_radix(16);
+        s.make_ascii_uppercase();
+        f.pad_integral(true, "0x", &s)
+    }
+}
+
+impl<const BIT_LEN: usize, const EXACT_LEN: bool> fmt::Binary
+    for FixedSizeBigInt<BIT_LEN, EXACT_LEN>
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.pad_integral(true, "0b", &self.0.to_str_radix(2))
+    }
+}
+
+impl<const BIT_LEN: usize, const EXACT_LEN: bool> fmt::Octal
+    for FixedSizeBigInt<BIT_LEN, EXACT_LEN>
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.pad_integral(true, "0o", &self.0.to_str_radix(8))
+    }
+}
+
+/// Helper macro for the `fixed_size_bigint` macro.
+macro_rules! fixed_size_bigint_impl {
+    ($struct_name:ident, $bit_len:expr, $exact_len:expr) => {
+        #[derive(Debug, Clone, Eq, PartialEq)]
+        pub struct $struct_name($crate::util::bigint::FixedSizeBigInt<$bit_len, $exact_len>);
+
+        // Use an unnamed free constant to avoid polluting the namespace where this is invoked.
```

Comment:
I feel like this is a well-known-enough trick that I wouldn't have included this caveat. =)

Reply:
Hehe, it was something new for me :)